Add bitmask parsing logic e.g., [&xx] [!&xx]

jrwdunham opened the pull request:
Fixes #117 

To run the tests:

    $ py.test -vs tests/test_prepare.py

Note: this also brings in the same change as https://github.com/openpreserve/fido/pull/120 so that PR 120 is a blocker on this PR.

jrwdunham commented:
@carlwilson do you know how I can get Codacy to stop complaining about my use of `assert` in the tests/test_prepare.py file?

ross-spencer commented:
Hi @jrwdunham  - while our permissions are in progress, i did a little code review. 

### Tests

1. We can wrap the complexity of the binrep call into a single function and then call the function in all the different tests. Just one way of making tests easier to write and refactor for folk.
```
def binrep_convert(byt):
    """Creates an eight-digit binary string for any given number between 0
    and 255. Values above 255 will not be further padded.
    """
    return bin(byt)[2:].zfill(8)
```

2. I think it would be more generally useful to have a test that sets up and tests more complicated sequences, something like:
```
def test_heterogenous_sequences():
    """Tests potential PRONOM sequences in their fullness, to monitor 
    syntactical components playing nicely with one other. 
    """
    pronom_bytesequence = 'ab{3}cd(01|02|03)~07ff'
    patt = convert_to_regex(pronom_bytesequence)
    assert re.search(patt, "ab010203cd0105ff")
    assert not re.search(patt, "ab010203cd0100ff")
```
That being said, I can't get these asserts to work... I might have over-simplified what is needed here or got something wrong but happy to work with you on the more general idea. The test can be set up in lists of tuples with sequences, test values, and expected outcomes.

### Main components in the code

1. `do_op_bitmasks()` might be a bit clearer with a fuller name. E.g. 'do_any_all_op_bitmasks()`. It can also be placed logically in the code below the two calling functions so that we arrive at it following reading those two.

2. For the compound statements such as you are using to generate the string. I think the compound statement might be the nicest way to do this, but PEP8 does have its issues with it (and for me they still take a little more time to read). I think a good compromise might be to use expanded variable names, e.g.

* `op` in the function argument can become `operand_type`
* `byt_` can become `byte` (it's a little close to our other use of `byt` at the moment

Overall I like your approach to this, especially the way you setup the test functions. 

If it helps, I will attach these comments inline once we have permissions to do CR that way.

jrwdunham commented:
@ross-spencer thanks for the CR!

I think I've implemented most of your suggestions in 21681c5a13adcbd22d7342e22caeeb692602e03a.

- Added `binrep_convert` function, but modified your docstring so pep257 wouldn't complain.
- Added tests for your heterogeneous sequences. More could be added too. Hopefully the new tests make it clear what kinds of input strings match the byte sequence you provided. If it's still unclear, it would probably be educational to hash it out here.
- Made the naming and ordering changes you suggested. Wasn't quite sure what you meant by "compound statements" so let me know if you still see possible improvements/clarifications.
- Used pytest's `@parametrize` to DRY-ify the tests

mistydemeo commented:
> do you know how I can get Codacy to stop complaining about my use of assert in the tests/test_prepare.py file?

I'm missing admin permissions in Codacy to configure this, but the documentation for this is here: https://support.codacy.com/hc/en-us/articles/207279979-Issues#2-remove-pattern
